session-manager/src/session.rs

```diff
@@ -12,63 +12,130 @@
 //! information on licensing and copyright.
 
 use crate::error::SessionManagerError;
+use mbedtls::alloc::List;
 use policy_utils::principal::Identity;
 
 use std::{
     io::{Read, Write},
+    sync::{Arc, Mutex},
     vec::Vec,
 };
 
-use rustls::{Certificate, ServerConnection};
-
 ////////////////////////////////////////////////////////////////////////////////
 // Sessions.
 ////////////////////////////////////////////////////////////////////////////////
 
 /// A principal is an individual identified with a cryptographic certificate,
 /// and assigned a set of roles that dictate what that principal can and cannot
 /// do in a Veracruz computation.
-pub type Principal = Identity<Certificate>;
+pub type Principal = Identity<List<mbedtls::x509::Certificate>>;
 
 /// A session consists of a TLS server session with a list of principals and
 /// their identifying information.
 pub struct Session {
     /// The TLS server session.
-    tls_connection: ServerConnection,
+    tls_context: mbedtls::ssl::Context<InsecureConnection>,
+    /// Whether the connection is established.
+    established: bool,
+    /// Read and write buffers shared with InsecureConnection.
+    shared_buffers: Arc<Mutex<Buffers>>,
     /// The list of principals, their identities, and roles in the Veracruz
     /// computation.
     principals: Vec<Principal>,
 }
 
+struct Buffers {
+    /// Read buffer used by mbedtls for cyphertext.
+    read_buffer: Vec<u8>,
+    /// Write buffer used by mbedtls for cyphertext.
+    write_buffer: Option<Vec<u8>>,
+}
+
+struct InsecureConnection {
```

Comment:
What's the purpose of the `InsecureConnection` struct?

Reply:
I'm unsure of the correct Rust terminology here but I'll add a comment like this:

Reply:
No, I see that another file already has a better explanation, which I can adapt:
```
/// This is the structure given to mbedtls and used for reading and
/// writing cyphertext, using the standard Read and Write traits.
```